src/prefect/records/filesystem.py

```diff
@@ -0,0 +1,161 @@
+        return TransactionRecord(
+            key=key, result=BaseResult.model_validate_json(record_data)
+        )
+
+    def write(self, key: str, result: BaseResult, holder: Optional[str] = None) -> None:
+        self._ensure_records_directory_exists()
+
+        if self.is_locked(key) and not self.is_lock_holder(key, holder):
+            raise ValueError(
+                f"Cannot write to transaction with key {key} because it is locked by another holder."
+            )
+
+        record_path = self.records_directory.joinpath(key)
+        record_path.touch(exist_ok=True)
+        record_data = result.model_dump_json()
+        record_path.write_text(record_data)
+
+    def exists(self, key: str) -> bool:
+        return self.records_directory.joinpath(key).exists()
+
+    def supports_isolation_level(self, isolation_level: IsolationLevel) -> bool:
+        return isolation_level in {
+            IsolationLevel.READ_COMMITTED,
+            IsolationLevel.SERIALIZABLE,
+        }
+
+    def acquire_lock(
+        self,
+        key: str,
+        holder: Optional[str] = None,
+        acquire_timeout: Optional[float] = None,
+        hold_timeout: Optional[float] = None,
+    ) -> bool:
+        holder = holder or self.generate_default_holder()
+
+        self._ensure_records_directory_exists()
+        lock_path = self._lock_path_for_key(key)
+
+        if self.is_locked(key) and not self.is_lock_holder(key, holder):
+            lock_free = self.wait_for_lock(key, acquire_timeout)
+            if not lock_free:
+                return False
+
+        try:
+            Path(lock_path).touch(exist_ok=False)
+        except FileExistsError:
+            if not self.is_lock_holder(key, holder):
+                logger.debug(
+                    f"Another actor acquired the lock for record with key {key}. Trying again."
+                )
+                return self.acquire_lock(key, holder, acquire_timeout, hold_timeout)
+
+        with open(Path(lock_path), "w") as lock_file:
+            lock_info = {"holder": holder}
+            if hold_timeout:
+                lock_info["expiration"] = str(
+                    pendulum.now("utc") + pendulum.duration(seconds=hold_timeout)
+                )
+            json.dump(lock_info, lock_file)
+
+        return True
+
+    def release_lock(self, key: str, holder: Optional[str] = None) -> None:
+        holder = holder or self.generate_default_holder()
+        lock_path = self._lock_path_for_key(key)
+        if not self.is_locked(key):
+            ValueError(f"No lock for transaction with key {key}")
+        if self.is_lock_holder(key, holder):
+            Path(lock_path).unlink()
+        else:
+            raise ValueError(f"No lock held by {holder} for transaction with key {key}")
+
+    def is_locked(self, key: str) -> bool:
+        lock_path = self._lock_path_for_key(key)
+        try:
+            with open(Path(lock_path)) as lock_file:
```

Comment:
This method is called in a lot of places and reads from the file each time; I think we can assume the key is locked if we find it's locked one time, and allow a caller to force reload through a kwarg if they want to truly recheck the file source of truth.

Reply:
Makes sense! I added a cache to the class and defaulted to using the cache where possible. I think the only part where I had to opt out of using the cache is in `wait_for_lock`, so the number of disk reads should be lower now.